database/database.go

```diff
@@ -1,127 +1,585 @@
 	if err != nil {
-		database.Close()
-		log.Fatal("Cannot create teachers table: ", err)
+		postgresDatabase.Close()
+		return errors.New("At Setup: " + err.Error())
 	}
 
-	_, err = database.Exec("CREATE TABLE IF NOT EXISTS quotes (id serial PRIMARY KEY, teacherid integer REFERENCES teachers (id), text varchar)")
+	// Create teachers table in PostgreSQL database if it doesn't exist
+	// for more information see TeachersT declaration
+	_, err = postgresDatabase.Exec(
+		`CREATE TABLE IF NOT EXISTS teachers (
+		TeacherID serial PRIMARY KEY, 
+		Name varchar, 
+		Title varchar, 
+		Note varchar)`)
 	if err != nil {
-		database.Close()
-		log.Fatal("Cannot create quotes table: ", err)
+		postgresDatabase.Close()
+		return errors.New("At Setup: " + err.Error())
 	}
-}
 
-// StoreQuote is a function to store quotes
-func StoreQuote(quote string, teacherid int) error {
-	database.Ping()
+	// Create quotes table in PostgreSQL database if it doesn't exist
+	// for more information see QuoteT declaration
+	_, err = postgresDatabase.Exec(
+		`CREATE TABLE IF NOT EXISTS quotes (
+		QuoteID serial PRIMARY KEY,
+		TeacherID integer REFERENCES teachers (TeacherID), 
+		Context varchar,
+		Text varchar,
+		Unixtime bigint,
+		Upvotes integer)`)
+	if err != nil {
+		postgresDatabase.Close()
+		return errors.New("At Setup: " + err.Error())
+	}
 
-	_, err := database.Exec("INSERT INTO quotes (teacherid, text) VALUES ($1, $2)", teacherid, quote)
+	// Create unverifiedQuotes table in PostgreSQL database if it doesn't exist
+	// for more information see UnverifiedQuoteT declaration
+	_, err = postgresDatabase.Exec(
+		`CREATE TABLE IF NOT EXISTS unverifiedQuotes (
+		QuoteID serial PRIMARY KEY,
+		Teacher varchar, 
+		Context varchar,
+		Text varchar,
+		Unictime bigint,
+		IPHash bigint)`)
 	if err != nil {
-		log.Print("Cannot store quote: ", err)
-		return err
+		postgresDatabase.Close()
+		return errors.New("At Setup: " + err.Error())
 	}
 
+	localDatabase.mux.Setup()
+	loadLocalDatabase()
+
 	return nil
+
 }
 
-// StoreTeacher is a function to store teachers
-func StoreTeacher(name string, title string, note string) error {
-	database.Ping()
+// Close database backend
+func Close() {
+	localDatabase.mux.LockWrite()
+	localDatabase.quoteSlice = nil
+	localDatabase.teacherSlice = nil
+	localDatabase.wordsMap = nil
+	localDatabase.mux.UnlockWrite()
+	postgresDatabase.Close()
+}
+
+/* -------------------------------------------------------------------------- */
+/*                          EXPORTED QUOTES FUNCTIONS                         */
+/* -------------------------------------------------------------------------- */
+
+// GetQuotes returns a slice containing all quotes
+// The weight variable will be zero
+func GetQuotes() *[]QuoteT {
+	localDatabase.mux.LockRead()
+	defer localDatabase.mux.UnlockRead()
+	quoteSlice := localDatabase.quoteSlice
+
+	return &quoteSlice
+}
+
+// GetQuotesByString returns a slice containing all quotes
```

Comment:
suggestion: only return those quotes that match at all / specify threshold (top 20?) for performance reasons

Reply:
- [ ] create a new function that sorts & filters - will be needed anyways

Reply:
described in #3 (REST API)